First

DmytroBylkov opened the pull request:
Додав магічні методи __eq__, __float__, __repr__ до класу Product
До класу ShoppingCart додав магычны методи __init__, __len__, __float__,__str__, __add__
Невийшло реалізувати методи, що в кошику б при додавання того ж товару змінював би кількість наявного, та видалення продукту з кошику. Можливо краще використовувати не список, для продуктів у кошику, а щось інше?

DmytroBylkov committed:
added methot __eq__ to class Product

DmytroBylkov committed:
added dunder metods float and str to Product and flaot to ShoppingCart

DmytroBylkov committed:
added method in class Shopping Cart

shorodilov commented on main.py:
```diff
@@ -23,14 +23,28 @@ class Product:
+    def __eq__(self, other):
+        """compares two products for identity"""
+
+        return self.name == other.name and self.price == other.price
```
Це спрацює лише для порівняння об'єктів класу `Product`.
Сюди треба додати перевірку типів. Можна просто повертати `False`, якщо тип не збігається.

shorodilov commented on main.py:
```diff
@@ -39,28 +53,53 @@ class ShoppingCart:
     """
-    products_in_cart = []
+    def __init__(self):
+        products_in_cart = []
```
Нащо ці рухи туди-сюди?
```suggestion
        self.products = []
```

shorodilov commented on main.py:
```diff
@@ -39,28 +53,53 @@ class ShoppingCart:
+        return self.total_price()
+
+    def __str__(self):
+        return self.products_in_cart
```
Метод `__str__` (як і `__repr__`) повинен повертати рядок.

shorodilov commented on main.py:
```diff
@@ -39,28 +53,53 @@ class ShoppingCart:
         sum = 0
         for product, quantity in self.products_in_cart:
             sum = sum + product.get_total(quantity)
         return sum
```
Не забувай про скруглення

shorodilov commented on main.py:
```diff
@@ -39,28 +53,53 @@ class ShoppingCart:
+        return self.products_in_cart
+
+    def __add__(self, other):
+        self.products_in_cart = self.products_in_cart + other.products_in_cart
```
Ліпше створюй новий кошик, не треба змінювати поточний.
І ще треба додати поведінку для випадку `cart_obj + product_obj`

shorodilov reviewed (changes requested):
Потребує змін

DmytroBylkov committed:
Update main.py

Refactoring __str__ method in Shopping cart, cnaged products list

shorodilov commented on main.py:
```diff
@@ -54,30 +54,30 @@ class ShoppingCart:
-        for product, quantity in self.products_in_cart:
+        for product, quantity in self.products:
             sum = sum + product.get_total(quantity)
         return sum
```
Не забувай про округлення.

DmytroBylkov replied:
Округлення в методі класу Product
def get_total(self, quantity: float):
        """calculating total price for product"""

        return round(quantity * self.price, 2)

shorodilov replied:
`0.1 + 0.2`???

DmytroBylkov committed:
Update main.py

Changed __eq__ in Product and add __eq__ in ShoppingCart, use __eq__ in def __add__ in ShoppingCart

DmytroBylkov commented:
Добрий день! Вніс правки за рекомендаціями. Лише округлення не вводив, так як воно відбувається в методі qet_total класу  Product. Та із створенням нового кошику не догнав ще, бо як тоді прописати щоб клас тепер не кошик products використовував у подальшому, а інший?

shorodilov commented on main.py:
```diff
@@ -39,28 +54,64 @@ class ShoppingCart:
+        """compares two products for identity"""
+        if not isinstance(other, ShoppingCart):
+            return False
+        return True
```
У цьому разі будь який кошик еквівалент будь якому. Треба додати логіку порівняння.

shorodilov commented on main.py:
```diff
@@ -39,28 +54,64 @@ class ShoppingCart:
+        return self.total_price()
+
+    def __str__(self):
+        return "".join(f"{product.__repr__()}: {quantity}, " for product, quantity in self.products)
```
```python
", ".join(...)
```

DmytroBylkov replied:
виправив

shorodilov commented on main.py:
```diff
@@ -39,28 +54,64 @@ class ShoppingCart:
+            self.products = self.products + other.products
+            return self.products
+        else:
+            return print("You can add only cart to cart")         
```
Що повертає `print`?

DmytroBylkov replied:
Прінт повертає в консоль коли не відбулося додавання обєкту, так як обєкт не пройшов порівняння

shorodilov replied:
`print` нічого не повертає.
Отже - що повертає твій метод?

shorodilov commented on main.py:
```diff
@@ -39,28 +54,64 @@ class ShoppingCart:
+        return "".join(f"{product.__repr__()}: {quantity}, " for product, quantity in self.products)
+
+    def __add__(self, other):
+        if self.__eq__(other):
```
Що робить оцей рядок?

DmytroBylkov replied:
перевіряє чи додається обєкт класу кошик, та якщо істина проходить додавання, інакше повертає print

shorodilov replied:
Не вірна відповідь. Він перевіряє `self == other`.
Навіщо це тут?

DmytroBylkov replied:
Я подумав, що якщо є метод перевірки, то при додаванні обєктів якраз і треба перевіряти чи додаються обєкти однакового класу

shorodilov replied:
Це все ще не виправлено.
Я бажаю поведінку `new_cart = cart_obj1 + cart_obj2` та `new_cart = cart_obj + product_obj`

shorodilov replied:
> Я подумав, що якщо є метод перевірки, то при додаванні обєктів якраз і треба перевіряти чи додаються обєкти однакового класу

То все вірно ти подумав.
Але треба саме таку поведінку прописати, а не використати якийсь рандомний метод.

shorodilov reviewed (changes requested):
Потребує виправлення

shorodilov commented:
> Добрий день! Вніс правки за рекомендаціями. Лише округлення не вводив, так як воно відбувається в методі qet_total класу  Product. Та із створенням нового кошику не догнав ще, бо як тоді прописати щоб клас тепер не кошик products використовував у подальшому, а інший?

Не зрозумів питання

shorodilov commented on main.py:
```diff
@@ -54,30 +54,30 @@ class ShoppingCart:
-        return self.products_in_cart
+        return "".join(f"{product.__repr__()}: {quantity}, " for product, quantity in self.products)
 
     def __add__(self, other):
```
```python
cart_obj + 1
```
???

DmytroBylkov replied:
додав вже опис що має бути обєкт класу ShoppingCart

shorodilov reviewed (changes requested):
Потребує виправлень

DmytroBylkov committed:
Update main.py

changed "" -> ", "

DmytroBylkov committed:
Update main.py

додав опис до аргументу в методі __add__

DmytroBylkov committed:
Update main.py

changed __eq__ in ShoppingCart

shorodilov commented on main.py:
```diff
@@ -39,28 +54,64 @@ class ShoppingCart:
     """
-    products_in_cart = []
+    def __init__(self):
+        products = []
```
Навіщо?

DmytroBylkov replied:
Та в умові завдання було додати конструктори до класів... До ShoppingCart можна не робити?

shorodilov replied:
Навіщо саме так.
У чому різниця з
```python
self.products = []
```

shorodilov replied:
???

shorodilov reviewed (changes requested):
Потребує виправлень

DmytroBylkov committed:
Update main.py

Змінив метод додавання, тепер він повертає список обєкту

shorodilov reviewed (changes requested):
Потребує змін

DmytroBylkov committed:
Update main.py

Переробив метод add, він додає продукт просто і продукти з іншого кошику.

shorodilov commented on main.py:
```diff
@@ -23,14 +23,32 @@ class Product:
+        return 'Product'
+      
+    def __str__(self) -> str:
+        return f'{self.name}'        
```
Це робити не обов'язково, у тебе і без того `self.name: str`.
Не помилка, але вводить в оману - треба буде виправити якось.

shorodilov commented on main.py:
```diff
@@ -39,28 +57,65 @@ class ShoppingCart:
+        return 'ShoppingCart'   
+   
+    def __add__(self, other, quantity=1):
+        if other.__repr__() == 'Product':
```
Користуйся `isinstance` для перевірки класу об'єкту.

shorodilov commented on main.py:
```diff
@@ -39,28 +57,65 @@ class ShoppingCart:
+        """compares two products for identity"""
+        if not isinstance(other, ShoppingCart):
+            return False
+        return self.__str__() == other.__str__()
```
```suggestion
        return set(self.products) == set(other.products)
```

shorodilov reviewed (changes requested):
Потребує допрацювання.
Загальне зауваження, щодо організації коду.
Методи ліпше впорядкувавати у межах класу: спочатку дандери, потім проперті, методи класу та статичні, потім методи інстансів, бо інакше важко шукати. От я не одразу знайшов `ShoppingCart.__str__` і навіть написав коментар, що за його відсутності перерівка кошиків на `==` буде йти інакше.

DmytroBylkov committed:
Update main.py

Додав метод hash у клас Product, для можливості порівнювати кошики через set

DmytroBylkov committed:
Update main.py

замінив перевірку в методі __add__

shorodilov reviewed (dismissed):
LGTM

shorodilov commented on main.py:
```diff
@@ -39,28 +60,65 @@ class ShoppingCart:
+    def __repr__(self) -> str:
+        return 'ShoppingCart'   
+   
+    def __add__(self, other, quantity=1):
```
```suggestion
    def __add__(self, other):
```

shorodilov replied:
Додай окремий метод для того, щоб складати товар до кошика.
Звідси просто виклакай його з quantity=1 якщо `cart_obj + product_obj`

DmytroBylkov committed:
Update main.py

Переробив метод __add__, проте ще не робоча версія так як при спрацюванні методу видає помилку як при додованні кошику, так і продукту
for product, quantity in self.products:
TypeError: cannot unpack non-iterable Product object

shorodilov reviewed (approved):
Начебто тепер все ок

DmytroBylkov committed:
Update main.py

shorodilov reviewed (approved):
lgtm
